### PR TITLE
Refactor assertGivesWarning into separate methods

### PR DESCRIPTION
**Pull request by djkirkham**
It's confusing that `assertGivesWarning` can be used to assert that no warning is raised. I've refactored the code into separate methods.

### CODE REVIEW

File: lib/iris/tests/unit/fileformats/netcdf/test_Saver.py [hunk 1]
```diff
@@ -440,7 +440,7 @@ def test_contains_default_fill_value_byte(self):
         # value if no fill_value argument is supplied when the data is of a
         # byte type.
         cube = self._make_cube('>i1')
-        with self.assertGivesWarning(r'\(fill\|mask\)', expect_warning=False):
+        with self.assertDoesNotGiveWarning(r'\(fill\|mask\)'):
```
Comment (pelson) [inline lib/iris/tests/unit/fileformats/netcdf/test_Saver.py]: rename? ```assertNoWarningsMatching```?

File: lib/iris/tests/__init__.py [hunk 2]
```diff
@@ -539,26 +539,37 @@ def assertRaisesRegexp(self, *args, **kwargs):
...
-        else:
-            if warning_raised:
-                msg = "Unexpected warning(s) raised, matching '{}' : {!r}."
-                msg = msg.format(expected_regexp, matches)
-                self.assertEqual(expect_warning, warning_raised, msg)
+        matches.extend(message for message in messages
+                        if expr.search(message))
+
+    @contextlib.contextmanager
+    def assertGivesWarning(self, expected_regexp=''):
```
Comment (pelson) [inline lib/iris/tests/__init__.py]: rename? ```assertWarningsMatching```?

Comment (pelson) [reply lib/iris/tests/__init__.py]: @djkirkham suggests: ```assertGivesWarningRegexp```

Comment (pelson) [reply lib/iris/tests/__init__.py]: Or even ```assertGivesWarningMatchingRegexp```

File: lib/iris/tests/unit/fileformats/netcdf/test_Saver.py (hunk 1 above)
Comment (pelson) [reply lib/iris/tests/unit/fileformats/netcdf/test_Saver.py]: @djkirkham suggests: ```assertDoesNotGiveWarningRegexp```

Comment (pelson) [reply lib/iris/tests/unit/fileformats/netcdf/test_Saver.py]: ```assertDoesNotGiveWarningMatchingRegexp```

Comment (pelson) [reply lib/iris/tests/unit/fileformats/netcdf/test_Saver.py]: ```assertNoWarningsMatchingRegexp```

**Comment by pp-mo** [comment]
On consideration, I don't like the original name.
But it shouldn't need to be as unwieldy as some of the recent suggestions.
Anyone for "AssertWarnsRegexp" and "AssertNoWarningsRegexp" ?

**Comment by pelson** [comment]
Some genuine test failures...

**Comment by pelson** [comment]
The failing test issue has been raised in #2875.

**Pull request merged**